staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -452,10 +453,14 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	// from oldest to newest
 	for _, d := range obj.(Deltas) {
 		switch d.Type {
-		case Sync, Added, Updated:
+		case Sync, Replaced, Added, Updated:
 			isSync := d.Type == Sync
 			s.cacheMutationDetector.AddObject(d.Object)
 			if old, exists, err := s.indexer.Get(d.Object); err == nil && exists {
+				// If the delta is from a relist, and the RV hasn't changed, elide it.
+				if d.Type == Replaced && isObjectSame(old, d.Object) {
+					continue
+				}
```

Comment:
is this fundamental to the fix, or a separate optimization? I expected just the change in how we compute `isSync`

Reply:
I would argue it's important. If a handler has requested no resyncs (i.e. no levels, just edges) we should presumably respect that.

Reply:
But, yes, we could set to isSync rather than eliding.

Reply:
> If a handler has requested no resyncs (i.e. no levels, just edges) we should presumably respect that.

That's not how I would have explained resyncs... I always took the expressed resync period to mean an artificial replay of all existing items in the store as update events, *not* the events you got on a relist which was required because the watch dropped and couldn't be restarted.

> we could alter isSync rather than eliding entirely

The PR has already altered isSync by introducing a new delta type... I thought that was sufficient to fix the observed bug

Reply:
Hmm, I see `processorListener#shouldResync` always returned false if a particular listener didn't request a resync, so those listeners would not have previously gotten any events from a relist (which got propagated as Sync delta events).

So, this currently changes the following behaviors:
* listeners that did not request a resync period now get notified of updated objects on relist (good)
* listeners that did request a resync period no longer get notified of unchanged objects on relist (questionable)

Reply:
> I see processorListener#shouldResync always returned false

Even worse, `sharedProcessor#shouldResync` is side-effecting! While not an issue per-se, side-effecting predicates are ... unexpected ...

> listeners that did request a resync period no longer get notified of unchanged objects on relist (questionable)

IMO, re-List / Replace is an interaction local to the reflector and its store (DeltaFIFO in this case). It should be DeltaFIFO's responsibility to generate the correct sequence of Added / Updated / Deleted / Sync deltas based on what it gets from the reflector. 

It doesn't seem valuable that SharedInformer handlers get a resync (of all objects) sooner than their ResyncInterval would suggest if the reflector does a re-List. It smells like a layer violation.